android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/RibCoroutineWorker.kt

```diff
@@ -0,0 +1,197 @@
+ * but with the additional [context] elements and [SupervisorJob], so a failure of any child launched in the scope won't
+ * fail the worker scope or its other children.
+ *
+ * Binding a worker is an asynchronous operation. To ensure [RibCoroutineWorker.onStart] is finished, callers can
+ * [await][BindWorkerHandle.await] or [join][BindWorkerHandle.join] the resulting [BindWorkerHandle] when in a coroutine.
+ *
+ * ```
+ * val handle = coroutineScope.bind(worker)
+ * handle.await() // wait for onStart to finish
+ * ```
+ *
+ * @receiver The parent scope of the supervisor scope passed on to [RibCoroutineWorker.onStart].
+ * @param worker The worker to be bound.
+ */
+@JvmOverloads
+public fun CoroutineScope.bind(
+  worker: RibCoroutineWorker,
+  context: CoroutineContext = RibDispatchers.Default,
+): BindWorkerHandle {
+  var deferred: CompletableDeferred<Unit>? = null
+  // Start the bind coroutine undispatched to make sure we set the deferred synchronously.
+  val bindJob = launch(context, start = CoroutineStart.UNDISPATCHED) {
+    supervisorScope {
+      deferred = CompletableDeferred<Unit>(coroutineContext.job).also {
+        // Cancel this binding job if deferred was cancelled.
+        // This deferred cancellation upon the job cancellation is already set through structured concurrency.
+        it.invokeOnCompletion { t -> if (t is CancellationException) cancel(t) }
+      }
+      // Verify cancellation and dispatch (if needed) using dispatcher in context instead of continuing on current thread.
+      dispatchIfNeeded()
+      worker.bindAndAwaitCancellation(deferred!!)
+    }
+  }
+  // We know deferred is non-null because binding coroutine was started undispatched.
+  return BindWorkerHandleImpl(deferred!!, bindJob)
+}
+
+private suspend inline fun dispatchIfNeeded() {
+  suspendCoroutineUninterceptedOrReturn sc@{ cont ->
+    val context = cont.context
+    val dispatcher = context[ContinuationInterceptor] as CoroutineDispatcher
+    if (!dispatcher.isDispatchNeeded(context)) return@sc Unit
+    // Coroutine was not in the right context -- we'll dispatch.
+    context.ensureActive()
+    cont.intercepted().resume(Unit)
+    COROUTINE_SUSPENDED
+  }
+  coroutineContext.ensureActive() // Don't continue if coroutine was cancelled after returning from dispatch.
+}
+
+@Suppress("TooGenericExceptionCaught") // Exception is not swallowed
+private suspend fun RibCoroutineWorker.bindAndAwaitCancellation(deferred: CompletableDeferred<Unit>) {
+  val terminationCause = try {
+    supervisorScope {
+      onStart(this)
+      ensureActive() // Before completing the binding deferred, make sure the binding coroutine is active.
+      deferred.complete(Unit)
+      awaitCancellation() // Never returns normally, so we are sure an exception will be caught.
+    }
+  } catch (t: Throwable) {
+    t
+  }
+  deferred.cancelOrCompleteExceptionally(terminationCause) // No effect if deferred is already completed.
+  onStop(terminationCause)
```

Comment:
Looks like we will never receive runtime crashes by this logic. I think will be really hard to catch the bug in large project, if your changes broking someone's else Worker. I would recommend to throw all Exceptions caused by `onStart` by default and introduce some `Cause` sealed class for cleaner handling : 
```kotlin
sealed class Cause {
  object Lifecycle : Cause()
  class Exception(val t: Throwable) : Cause()
}
```

Reply:
That's an interesting criticism, @SergeySmykovskyi. However, given that binding is an asynchronous operation in this design (`bind` is non-suspend, called outside of a coroutine), I don't see how we can synchronously throw! Exposing exceptions to the user in coroutines is done by `await`-ing in coroutines. 

To address this criticism, `bind` would need to be `suspend` itself (so it could throw into the caller coroutine), but it gives rise to another set of issues as discussed in the design -- mainly that it's poor practice to both launch asynchronous operations (non-terminating coroutines like shared flow collection) and suspend in a `suspend fun`. When a function is `suspend`, callers expect that by the time it returns it has already done all of its work. Alternatively, if we did not pass in a `CoroutineScope` on `onStart` (so users would need to open its own scope to collect non-terminating flow, using for example `supervisorScope { }`), the binding operation would never terminate. 

Do you have a concrete suggestion on how to tackle this?

Reply:
Might need deeper investigation but I saw it as simple solution. Default implementation of `onStop` :
```kotlin
  @JvmDefault
  @CallSuper // Required to be sure that even overridden `onStop`  will throw unhandled exception (something like in `RXJava`)
  public fun onStop(cause: Cause) {
    if (cause is Cause.Exception) {
      handleException(cause.t)
    }
  }

  @JvmDefault
  public fun handleException(t: Throwable) {
    throw t 
  }
```

Reply:
@SergeySmykovskyi I've taken your concern into consideration and decided to change the binding job from `async` (exposes exception to user by means of `.await()`) into `launch` (automatically propagates exception to parent).

This will cause the parent scope to fail with exception, and if parent is a supervisor without a `CoroutineExceptionHandler`, it will crash the application *after* delivering the exception to `onStop`.

Reply:
Validated latest changes on internal codebase. Throwing on `onStart` crashes the app. Unbinding (normal cancellation) doesn't.